### PR TITLE
[Mgmt] Azure.ResourceManager.MachineLearning: Migrate to TypeSpec (API version 2025-12-01)

### LINKED ISSUE #56943: ** MPG migration - machinelearningservices

### PR DESCRIPTION
**Pull request by live1206**
fixes #56943

### Description

Migrates `Azure.ResourceManager.MachineLearning` from Swagger/AutoRest to TypeSpec-based generation using the Azure Management Generator.

- **API version**: 2024-04 (Swagger) -> 2025-12-01 (TypeSpec)
- **Spec**: `specification/machinelearningservices/MachineLearningServices.Management`
- **Spec PR**: https://github.com/Azure/azure-rest-api-specs/pull/41372
- **Emitter**: `@azure-typespec/http-client-csharp-mgmt`

### Migration Status

- TypeSpec generation is wired through `tsp-location.yaml` and points to the latest MachineLearningServices spec PR commit.
- AutoRest generation is removed (`autorest.md` and `IncludeAutorestDependency`).
- Generated code has been refreshed from TypeSpec and customizations have been updated for the generated type names and operation shapes.
- 1.2.4 GA API compatibility mitigations are applied with targeted custom code and TypeSpec renames instead of broad ApiCompat suppressions.
- Remaining ApiCompat baseline suppressions are limited to the expected MPG WirePath migration pattern.

### Compatibility / Review Fixes

- Preserved legacy public model factory overloads and child-resource getter shapes where needed for GA compatibility.
- Restored `PrivateEndpointBase` and `RegistryPrivateEndpoint : PrivateEndpointBase` public shape.
- Removed the obsolete custom `RegistryPrivateEndpoint` serialization shim and let `RegistryPrivateEndpoint` generate from TypeSpec.
- Renamed generated public acronym type `MachineLearningExternalFQDNResult` to `MachineLearningExternalFqdnResult` through a scoped C# TypeSpec rename.
- Kept custom compatibility model/serialization code only where generation cannot preserve the previous GA surface directly.

### Spec Changes

- Added scoped `@@clientName(..., "csharp")` customizations for backward-compatible C# type names.
- Fixed duplicate operations and property names needed for SDK generation.
- Upgraded `armCommonTypesVersion` from v3 to v6.

### Validation

- `dotnet build`
- `pwsh ./eng/scripts/Export-API.ps1 machinelearningservices`
- Targeted validation tests on `net10.0`:
  - `ValidateInheritanceForResourceAndCollectionSuffix`
  - `ValidateModelReaderWriterPattern`

### CODE REVIEW

File: sdk/machinelearningservices/Azure.ResourceManager.MachineLearning/src/Customized/Models/UnknownAssetReferenceBase.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+#nullable disable
+
+using Microsoft.TypeSpec.Generator.Customizations;
+
+namespace Azure.ResourceManager.MachineLearning.Models
+{
+    // Customized: this generated unknown fallback type is not a TypeSpec symbol, so @@clientName cannot target it.
+    [CodeGenType("UnknownMachineLearningAssetReferenceBase")]
+    internal partial class UnknownAssetReferenceBase { }
```
Comment (live1206) [inline sdk/machinelearningservices/Azure.ResourceManager.MachineLearning/src/Customized/Models/UnknownAssetReferenceBase.cs]: since this is internal, if this only affects the attribute, we can consider to add them to baseline file instead of customizing internal models.

**Review by live1206 — COMMENTED** [review]
## MPG migration review

**Review disposition:** These findings are blocking and would normally be submitted as `REQUEST_CHANGES`, but GitHub does not allow this account to request changes on its own pull request; submitting as `COMMENT` instead.

### Phase 1 - Versioning
Blocking: `ApiCompatVersion` was downgraded from `1.2.4` to `1.2.3`, which would hide compatibility checks against the latest stable package. The PR also removes the `1.2.4` changelog entry from `main`. The apicompat baseline contains targeted WirePath removals, which are allowed for MPG migration, but it also adds a `CannotChangeAttribute` entry, which should be fixed by preserving the API/wire metadata rather than suppressing the break. No major version bump was found.

### Phase 2 - API review
The trusted naming scanner, using `Azure.ResourceManager.MachineLearning_1.2.4` as the baseline, reported 48 new naming/type-shape violations after filtering existing baseline issues. Contextual naming: evaluated 164 new public types, flagged 3 deterministic resource-name/drift issues (`Connection*`/`WorkspaceResource`), plus the scanner-reported generic/resource infix/suffix issues. Key blocking examples are commented inline; remaining scanner issues should be resolved with scoped TypeSpec customizations where possible (`@@clientName(..., "csharp")`, `@@alternateType(..., "csharp")`) rather than SDK shims.

### Phase 3 - Breaking change detection
Blocking: comparing against the recovered 1.2.4 API shows 63 removed/changed public API lines, including extension methods, collection `GetAll` overloads, ModelFactory overloads, and model constructors/properties. The downgraded `ApiCompatVersion` would mask these. Restore `ApiCompatVersion` to `1.2.4`, remove unsupported baseline entries, and mitigate real breaks through TypeSpec decorators or compatibility customizations.

### Phase 4 - Migration customization review
- 4.1: Some customizations still lack enough root-cause justification; the throwing workspace-connection compatibility methods need an explanation of why they cannot delegate safely.
- 4.2: `CodeGenMember` usage looked appropriate in sampled same-data rename cases.
- 4.3: No redundant `CodeGenType` usage found.
- 4.4/4.5: `WorkspaceResource` is a new generic compatibility shape instead of restoring/preserving the shipped `MachineLearningWorkspaceResource` surface.
- 4.6: `LegacyModelFactoryOverloads` has 12 self-recursive overloads that will stack overflow.
- 4.7: No constructor-suppression blocker was selected beyond the API removals called out above.
- 4.8: No using-only customization diff selected.
- 4.9: The workspace-connection compatibility methods throw without demonstrating that old behavior cannot be implemented.
- 4.10: No public method input parameter typed as `IList<T>` was found in custom code; the `IList<T>` hits were properties/private helpers.
- 4.11: The project excludes a generated file from compilation as a generator-bug workaround; fix the generator/TypeSpec hierarchy and regenerate instead of carrying a generated-output exclusion.

### Phase 5 - TypeSpec decorator preference
Prefer scoped TypeSpec fixes for the generated API-shape issues: `@@clientName(..., "csharp")` for generic resource/model names such as `ConnectionResource`, `WorkspaceResource`, and `*Request`/`*Response` models; `@@alternateType(..., "csharp")` for ARM resource IDs and Azure locations; and hierarchy/resource-shape fixes for the generated file exclusion. No `client.tsp` changes are present in this repo diff to target inline.

### Phase 6 - Migration new API triage
Accepted as likely real API-version additions: many new capability host, inference endpoint/group/pool, RAI, serverless, fine-tuning, quota, and vulnerability types. Flagged rename/drift count: 3 resource-name/drift issues plus 48 scanner findings. Duplicate model/type count: 1 clear duplicate/generic workspace resource shape; duplicate enum count: 0 selected; duplicate operation/resource-shape count: 1 selected (`ConnectionResource` deployment shape).

Final counts: critical/blocking issues: 7 inline groups; should-fix issues: 48 scanner findings and 63 API removals/changes to triage/mitigate; suggestions: prefer scoped TypeSpec decorators over SDK shims. Total inline comment count: 6.

**Review by live1206 — COMMENTED** [review]
## MPG migration review (current diff)

**Review disposition:** Blocking issues remain. I am submitting this as `COMMENT` because this account appears to be the PR author and GitHub does not allow requesting changes on your own PR; treat the findings below as `REQUEST_CHANGES` severity.

### Phase 1 - Versioning
Blocking: the project still has `ApiCompatVersion` set to `1.2.3`, but the latest stable tag for this package is `Azure.ResourceManager.MachineLearning_1.2.4`. Please update ApiCompat to `1.2.4` so compatibility is checked against the latest shipped API. No major version bump was found. The new apicompat baseline contains only targeted `WirePathAttribute` removal entries, which are allowed for MPG migration.

### Phase 2 - API review
The trusted naming scanner was run against the PR API file using `Azure.ResourceManager.MachineLearning_1.2.4` as baseline. It filtered 50 baseline issues and reported 44 current new issues: 25 errors and 19 warnings. Inline comments cover representative deterministic violations that should be fixed with scoped TypeSpec customization where possible; the remaining scanner findings are non-inline below.

Contextual naming: evaluated 163 new public types, flagged 14 scanner/contextual resource/type-shape groups, including `EndpointResourcePropertiesBasicResource*`, `ManagedNetworkSettingsPropertiesBasicResource*`, `RaiBlocklist*PropertiesBasicResource*`, `RaiPolicyPropertiesBasicResourceData`, `ErrorResponse`, `WorkspacePropertiesUpdateParameters`, request/response suffix models, all-caps acronym names, and generic `PropertiesBase`/`RequestConfiguration`-style names.

### Phase 3 - Breaking change detection
Comparing the PR API against the 1.2.4 API listing did not show removed public API lines in the listing diff. This phase still depends on restoring `ApiCompatVersion` to `1.2.4`; otherwise CI can miss compatibility issues against the latest stable package.

### Phase 4 - Migration customization review
- 4.1: The sampled customization files now include root-cause comments for the legacy shims/suppressions.
- 4.2: Sampled same-data rename/alias customizations use `CodeGenMember` appropriately.
- 4.3: No new redundant `CodeGenType` blocker found in sampled files.
- 4.4/4.5: The earlier duplicate `WorkspaceResource` issue appears addressed; `MachineLearningWorkspaceConnectionDeployment*` is now contextual.
- 4.6: The earlier recursive ModelFactory overload issue appears fixed; sampled overloads now construct/translate generated models directly.
- 4.7: No constructor-suppression blocker selected beyond scanner/type-shape issues.
- 4.8: No using-only customization diff selected.
- 4.9: The legacy workspace-connection methods still throw, but now document why the shipped PUT shape cannot be replayed safely and point to the PATCH replacement.
- 4.10: No externally visible public method input parameter typed as `IList<T>` was found in custom code; public custom methods use `IEnumerable<T>` where relevant.
- 4.11: The earlier generated-file `Compile Remove` workaround appears removed.

### Phase 5 - TypeSpec decorator preference
Prefer scoped TypeSpec fixes over SDK shims for the remaining generated surface issues: `@@clientName(..., "csharp")` for suffix/acronym/contextual names, `@@alternateType(..., Azure.Core.ResourceIdentifier, "csharp")` for ARM resource IDs, `@@alternateType(..., Azure.Core.AzureLocation, "csharp")` for Azure locations, and `@@access(..., Access.internal, "csharp")`/common error handling for public error-response wrappers that should not be exposed.

### Phase 6 - Migration new API triage
Accepted as likely real API-version additions: capability host, inference endpoint/group/pool, RAI/blocklist/policy, fine-tuning, quota, vulnerability, and serverless-related surfaces. Flagged rename/drift count: 14 naming/type-shape groups from the scanner/contextual pass. Duplicate model/type count: 1 ARM-common duplicate (`ErrorResponse`); duplicate enum count: 0 selected; duplicate operation/resource-shape count: 0 selected after the latest workspace-connection-deployment rename.

### Non-inline findings
Remaining scanner issues not individually commented inline: `ServerlessInferenceEndpointAuthMode.AAD`/`KeyAndAAD`; `FQDNEndpointsPropertyBag`; `PTUDeploymentUsage`; boolean property names `ProvisionNetworkNow`, `DynamicThrottlingEnabled`, `Exists`, `ReleaseQuotaOnStop`, `PremiumIO`, `CreateHostPath`, `RebootPending`, `Blocking`, `IncludeDashboard`, `ExpirableSecret`, `ServerlessComputeNoPublicIP`, `Patchable`; DateTimeOffset names `FineTune` and `Inference`; `EndpointResourceProperties.Location` and `RegistryListCredentialsResult.Location` as strings; request/response suffix types `DeltaModelListRequest`, `DeltaModelModifyRequest`, `DeltaModelStatusRequest`, `DeltaModelStatusResponse`; and the related `Resource` infix generated data/collection types for endpoint, managed network settings, RAI blocklist item/blocklist, and RAI policy.

Final counts: critical/blocking issues: 7 inline groups plus the non-inline ApiCompatVersion issue; should-fix issues: 44 scanner findings total; suggestions: prefer scoped TypeSpec decorators over SDK custom shims for generated naming/type fixes. Total inline comment count: 6.

**Review by live1206 — COMMENTED** [review]
Focused duplicate API scan for MPG migration (same pattern checked in SecurityCenter):

- Compared current API against the latest shipped `Azure.ResourceManager.MachineLearning_1.2.4` API.
- Scanned `src/Customized/**` for explicit hide-and-replace shims (`renamed GA model to ...`, `ToGenerated`, generated replacement delegation). **No current explicit renamed-GA custom shim was found**.
- Scanned current/new API pairs for old+new public types (`Foo` + `FooInfo`/`FooData`/`FooResult`/`FooPropertiesBasicResource*`) and exact duplicate public shapes.

**Result:** no new unreported true duplicate/hide-and-replace API finding on the current head.

Previously reported duplicate/drift items are already covered or fixed:
- `WorkspaceResource` duplicate shape: fixed per existing thread.
- `ConnectionResource` generic deployment shape: fixed per existing thread.
- `ErrorResponse` ARM-common duplicate: fixed per existing thread.
- Remaining `*PropertiesBasicResource*` / `*PropertiesBasicResourceData` items are still naming/resource-infix/contextual-shape issues already covered in prior review summaries, but I did not find evidence that they duplicate a shipped GA type with a hidden custom shim.

**Review by live1206 — COMMENTED** [review]
Posted inline comments for the remaining BasicResource naming/resource-shape findings so they can be tracked individually.

**Review by live1206 — COMMENTED** [review]
Posted inline comments for remaining current-head issues from the local re-review. I did not post the `IsDynamicThrottlingEnabled` TTL scanner warning because it appears to be a false positive: the property is a boolean flag, not a TTL/duration value.

**Review by live1206 — COMMENTED** [review]
Latest re-review for head a6b6539: most previous MachineLearning API comments are fixed. Remaining inline finding is the new `ExternalFQDNResponse` naming issue. I did not duplicate the existing tsp-location fork thread.

File: sdk/machinelearningservices/Azure.ResourceManager.MachineLearning/src/Generated/Models/MachineLearningExternalFQDNResult.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+#nullable disable
+
+using System;
+using System.Collections.Generic;
+using Azure.ResourceManager.MachineLearning;
+
+namespace Azure.ResourceManager.MachineLearning.Models
+{
+    /// <summary> The MachineLearningExternalFQDNResult. </summary>
+    public partial class MachineLearningExternalFQDNResult
```
Comment (live1206) [inline sdk/machinelearningservices/Azure.ResourceManager.MachineLearning/src/Generated/Models/MachineLearningExternalFQDNResult.cs]: **[Phase 2 / ACRONYM002]** `MachineLearningExternalFQDNResult` is a new public type and uses all-caps `FQDN`. .NET naming conventions require 3+ letter acronyms to be PascalCase, so this should be `MachineLearningExternalFqdnResult`. Since this is generated, please fix it through a scoped TypeSpec rename such as `@@clientName(..., "csharp")` and regenerate rather than editing generated code by hand.

File: eng/apicompatbaselines/Azure.ResourceManager.MachineLearning.txt [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+CannotRemoveAttribute : Attribute 'Azure.ResourceManager.MachineLearning.WirePathAttribute' exists on 'Azure.ResourceManager.MachineLearning.Models.EnvironmentVariable.AdditionalProperties' in the contract but not the implementation.
+CannotRemoveAttribute : Attribute 'Azure.ResourceManager.MachineLearning.WirePathAttribute' exists on 'Azure.ResourceManager.MachineLearning.Models.ImageSetting.AdditionalProperties' in the contract but not the implementation.
+
+# PrivateEndpointBase was a duplicate generated shape for MachineLearningPrivateEndpoint. The TypeSpec
+# customization now routes the shared private endpoint model to MachineLearningPrivateEndpoint.
+MembersMustExist : Member 'public Azure.ResourceManager.MachineLearning.Models.PrivateEndpointBase Azure.ResourceManager.MachineLearning.Models.ArmMachineLearningModelFactory.PrivateEndpointBase(Azure.Core.ResourceIdentifier)' does not exist in the implementation but it does exist in the contract.
+TypesMustExist : Type 'Azure.ResourceManager.MachineLearning.Models.PrivateEndpointBase' does not exist in the implementation but it does exist in the contract.
```
Comment (live1206) [inline eng/apicompatbaselines/Azure.ResourceManager.MachineLearning.txt]: **[Phase 1 / Phase 3]** These `PrivateEndpointBase`/`RegistryPrivateEndpoint` ApiCompat suppressions are not the MPG WirePath exception. `PrivateEndpointBase` and the `RegistryPrivateEndpoint : PrivateEndpointBase` inheritance existed in the 1.2.4 public API, so suppressing their removal hides a real breaking change. Please mitigate this with compatibility/customization code (or restore the previous public shape) instead of adding baseline suppressions. The WirePath entries above are the only baseline suppressions expected for this migration pattern.

**Review by live1206 — COMMENTED** [review]
Re-reviewed the latest push. Most earlier issues are fixed, but I found two new blocking issues in addition to the existing unresolved `tsp-location.yaml` fork reference thread.

**Review by ArthurMa1978 — APPROVED** [review]
